Code/Tools/AssetProcessor/Platform/Windows/native/FileWatcher/FileWatcher_win.cpp

```diff
@@ -6,125 +6,149 @@
-            m_thread = std::thread(std::bind(&FolderRootWatch::WatchFolderLoop, this));
-            return true;
-        }
+        return false;
     }
-    return false;
+
+    return folderWatch->second.ReadChanges();
 }
 
-void FolderRootWatch::Stop()
+bool FileWatcher::PlatformImplementation::FolderRootWatch::ReadChanges()
+{
+    // Register to get directory change notifications for our directory handle
+    return ::ReadDirectoryChangesW(
+        m_directoryHandle.get(),
+        &m_fileNotifyInformationList,
+        sizeof(m_fileNotifyInformationList),
+        m_recursive,
+        FILE_NOTIFY_CHANGE_LAST_WRITE | FILE_NOTIFY_CHANGE_DIR_NAME | FILE_NOTIFY_CHANGE_ATTRIBUTES | FILE_NOTIFY_CHANGE_FILE_NAME,
+        nullptr,
+        m_overlapped.get(),
+        nullptr
+    );
+}
+
+void FileWatcher::PlatformStop()
 {
     m_shutdownThreadSignal = true;
-    CloseHandle(m_platformImpl->m_ioHandle);
-    m_platformImpl->m_ioHandle = nullptr;
 
+    // Send a special signal to the child thread, that is blocked in a GetQueuedCompletionStatus call, with a pointer
+    // to this as the completion key.
+    PostQueuedCompletionStatus(m_platformImpl->m_ioHandle.get(), 0, reinterpret_cast<ULONG_PTR>(this), nullptr);
     if (m_thread.joinable())
     {
         m_thread.join(); // wait for the thread to finish
-        m_thread = std::thread(); //destroy
+        m_thread = AZStd::thread(); //destroy
     }
-    CloseHandle(m_platformImpl->m_directoryHandle);
-    m_platformImpl->m_directoryHandle = nullptr;
 }
 
-void FolderRootWatch::WatchFolderLoop()
+void FileWatcher::WatchFolderLoop()
 {
-    FILE_NOTIFY_INFORMATION aFileNotifyInformationList[50000];
-    QString path;
-    OVERLAPPED aOverlapped;
-    LPOVERLAPPED pOverlapped;
-    DWORD dwByteCount;
-    ULONG_PTR ulKey;
+    LPOVERLAPPED directoryId = nullptr;
+    ULONG_PTR completionKey = 0;
 
     while (!m_shutdownThreadSignal)
     {
-        ::memset(aFileNotifyInformationList, 0, sizeof(aFileNotifyInformationList));
-        ::memset(&aOverlapped, 0, sizeof(aOverlapped));
-
-        if (::ReadDirectoryChangesW(m_platformImpl->m_directoryHandle, aFileNotifyInformationList, sizeof(aFileNotifyInformationList), true, FILE_NOTIFY_CHANGE_LAST_WRITE | FILE_NOTIFY_CHANGE_DIR_NAME | FILE_NOTIFY_CHANGE_ATTRIBUTES | FILE_NOTIFY_CHANGE_FILE_NAME, nullptr, &aOverlapped, nullptr))
+        DWORD dwByteCount = 0;
+        if (::GetQueuedCompletionStatus(m_platformImpl->m_ioHandle.get(), &dwByteCount, &completionKey, &directoryId, INFINITE))
         {
-            //wait for up to a second for I/O to signal
-            dwByteCount = 0;
-            if (::GetQueuedCompletionStatus(m_platformImpl->m_ioHandle, &dwByteCount, &ulKey, &pOverlapped, INFINITE))
+            // For normal directory changed notifications, we're setting the completion key to 1 (see the call to
+            // CreateIoCompletionPort). In the case the thread needs to be shut down, a pointer to this is sent for the
```

Comment:
It's a little awkward that the pointer address is used as the sentinel value to indicate that this thread should shutdown (instead of some status code in an enum or something)

Reply:
I've made an enum to better describe what's happening here.